src/Adapter/Product/Combination/Create/CombinationCreator.php

```diff
@@ -125,54 +152,76 @@ private function formatScalarValues(array $groupedAttributeIdsList): array
     /**
      * @param Product $product
      * @param Traversable $generatedCombinations
+     * @param ShopId[] $shopIds
      *
-     * @return CombinationId[]
-     *
-     * @throws CoreException
+     * @return CombinationId[] the ids of newly created combinations
      */
-    private function addCombinations(Product $product, Traversable $generatedCombinations): array
+    private function addCombinations(Product $product, Traversable $generatedCombinations, array $shopIds): array
     {
         $product->setAvailableDate();
         $productId = new ProductId((int) $product->id);
-        $alreadyHasCombinations = $hasDefault = (bool) $this->combinationRepository->findDefaultCombination($productId);
+        $hasCombinations = $this->productRepository->hasCombinations($productId);
+        $newCombinationIds = [];
+
+        foreach ($shopIds as $shopId) {
+            foreach ($generatedCombinations as $generatedCombination) {
+                if (!$hasCombinations) {
+                    // Product has no combinations yet, so we create new combinations and skip the rest of the loop
+                    $newCombinationIds[] = $this->persistCombination($productId, $generatedCombination, $shopId);
+                    continue;
+                }
 
-        $addedCombinationIds = [];
-        foreach ($generatedCombinations as $generatedCombination) {
-            // Product already has combinations so we need to filter existing ones
-            if ($alreadyHasCombinations) {
                 $attributeIds = array_values($generatedCombination);
-                $matchingCombinations = $this->combinationRepository->getCombinationIdsByAttributes($productId, $attributeIds);
-                if (!empty($matchingCombinations)) {
+                $matchingCombinationId = $this->combinationRepository->findCombinationIdByAttributes($productId, $attributeIds);
+
+                if (!$matchingCombinationId) {
+                    // if there is no combination of provided attributes yet, we create a new one and skip to next iteration
+                    $newCombinationIds[] = $this->persistCombination($productId, $generatedCombination, $shopId);
+                    continue;
+                }
+
+                // if there is a combination of provided attributes, and it is already associated with the shop, then we don't do anything and skip to next iteration
+                if ($this->combinationRepository->isAssociatedWithShop($matchingCombinationId, $shopId)) {
                     continue;
                 }
+
+                // when combination already exists in product_attribute, then we add it to related product_attribute_shop
+                $this->combinationRepository->addToShop($matchingCombinationId, $shopId);
+                // create dedicated stock_available for combination in related shop
+                $this->stockAvailableMultiShopRepository->createStockAvailable($productId, $shopId, $matchingCombinationId);
```

Comment:
It's an excellent idea to create the StockAvailable on the combination creation. However I believe this is only done for new associations, it's not handled in `persistCombination` nor in `CombinationRepository::create` but I guess it should no? (Probably it's better to handle it in create directly)

Without this, we can have new combinations without associated StockAvailable, and it becomes a bit more complex with groups that share their stock. Since `createStockAvailable` already handles the complexity about group shared stock it's safer to use the sooner possible and make sure the data is consistent from the start.

Reply:
in Combination::add() the method StockAvailable::setProductOutOfStock is called and as I checked it creates the stock_available as a side effect, so its all good